### PR TITLE
Fix the debugger in case of a local shadows a var

### LINKED ISSUE #846: Name collision causes an exception during debugging
I use [di](https://github.com/darkleaf/di) in my project and there is a lot of code like this

```clojure
(ns debugger-example)

(defn service [-deps a b]
  [a b])

;; C-u C-M-x
(defn caller [{service  `service
               service* `service}]
  [(service* 1 2)
   (service  1 2)]) ;; => works without debugger [[1 2] [1 2]]

(comment
  (caller {`service (partial service :unused-deps)})
  ,,,)
```

Unfortunately, I can't use the debugger.

## Expected behavior

I can successfully execute `caller` with debugger mode.

## Actual behavior

It shows value of `(service* 1 2)` but fails on `(service 1 2)`.

Снимок экрана 2024-02-04 в 18 45 15

```
1. Unhandled clojure.lang.ArityException
   Wrong number of args (2) passed to: debugger-example/service

                  AFn.java:  429  clojure.lang.AFn/throwArity
                  AFn.java:   36  clojure.lang.AFn/invoke
                  AFn.java:  156  clojure.lang.AFn/applyToHelper
                  AFn.java:  144  clojure.lang.AFn/applyTo
                  Var.java:  707  clojure.lang.Var/applyTo
                  core.clj:  667  clojure.core/apply
                  core.clj:  662  clojure.core/apply
                 debug.clj:  542  cider.nrepl.middleware.debug/apply-instrumented-maybe
                 debug.clj:  538  cider.nrepl.middleware.debug/apply-instrumented-maybe
                      REPL:    7  debugger-example/eval23279/caller
                      REPL:   12  debugger-example/eval23283
                      REPL:   12  debugger-example/eval23283
             Compiler.java: 7177  clojure.lang.Compiler/eval
             Compiler.java: 7132  clojure.lang.Compiler/eval
                  core.clj: 3229  clojure.core/eval
                  core.clj: 3225  clojure.core/eval
    interruptible_eval.clj:   87  nrepl.middleware.interruptible-eval/evaluate/fn/fn
                  AFn.java:  152  clojure.lang.AFn/applyToHelper
                  AFn.java:  144  clojure.lang.AFn/applyTo
                  core.clj:  667  clojure.core/apply
                  core.clj: 1990  clojure.core/with-bindings*
                  core.clj: 1990  clojure.core/with-bindings*
               RestFn.java:  428  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   87  nrepl.middleware.interruptible-eval/evaluate/fn
                  main.clj:  437  clojure.main/repl/read-eval-print/fn
                  main.clj:  437  clojure.main/repl/read-eval-print
                  main.clj:  459  clojure.main/repl/fn
                  main.clj:  459  clojure.main/repl
                  main.clj:  368  clojure.main/repl
               RestFn.java: 1526  clojure.lang.RestFn/invoke
    interruptible_eval.clj:   84  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:   56  nrepl.middleware.interruptible-eval/evaluate
    interruptible_eval.clj:  152  nrepl.middleware.interruptible-eval/interruptible-eval/fn/fn
                  AFn.java:   22  clojure.lang.AFn/run
               session.clj:  218  nrepl.middleware.session/session-exec/main-loop/fn
               session.clj:  217  nrepl.middleware.session/session-exec/main-loop
                  AFn.java:   22  clojure.lang.AFn/run
               Thread.java: 1583  java.lang.Thread/run
```

### cider-nrepl version

+ cider-nrepl 0.45.0
+ CIDER 1.13.1 (Santiago)
+ nREPL 1.0.0

### Java version

21.0.1

### Operating system

mac os x 13.6.1

### PR DESCRIPTION
**Pull request by darkleaf**
Fixes #846

Before submitting a PR make sure the following things have been done:

- [x] The commits are consistent with our [contribution guidelines](../blob/master/.github/CONTRIBUTING.md)
- [x] You've added tests to cover your change(s)
- [x] All tests are passing
- [x] You've updated the README
- [x] Middleware documentation is up to date
  * Please check out and modify the `cider.nrepl` ns which has all middleware documentation.
  * Run `lein docs` afterwards, and commit the results.

**Note:** If you're just starting out to hack on `cider-nrepl` you might find
[nREPL's documentation](https://nrepl.org) and the
"Design" section of the README extremely useful.*

Thanks!

***

1. `resolve` have an extra arity for `&env`
2. I added arity for `looks-step-innable?` for backwards compatibility

```clojure
(defn ns-resolve
  "Returns the var or Class to which a symbol will be resolved in the
  namespace (unless found in the environment), else nil.  Note that
  if the symbol is fully qualified, the var/Class to which it resolves
  need not be present in the namespace."
  {:added "1.0"
   :static true}
  ([ns sym]
    (ns-resolve ns nil sym))
  ([ns env sym]
    (when-not (contains? env sym)
      (clojure.lang.Compiler/maybeResolveIn (the-ns ns) sym))))

(defn resolve
  "same as (ns-resolve *ns* symbol) or (ns-resolve *ns* &env symbol)"
  {:added "1.0"
   :static true}
  ([sym] (ns-resolve *ns* sym))
  ([env sym] (ns-resolve *ns* env sym)))
```

### CODE REVIEW

File: src/cider/nrepl/middleware/debug.clj [hunk 1]
```diff
@@ -448,16 +448,18 @@ this map (identified by a key), and will `dissoc` it afterwards."}
...
-      ;; at this point, which is at macroexpansion time.
-      (and v
-           (safe-to-debug? (:ns m))
-           (not (:macro m))
-           (not (:inline m))))))
+  ([form]
+   (looks-step-innable? nil form))
+  ([env form]
+   (when (and (seq? form) (symbol? (first form)))
+     (let [v (resolve env (first form)) ; Note: special forms resolve to nil
```
Comment (vemv) [inline src/cider/nrepl/middleware/debug.clj:455]: Note that there's another `resolve` in this file - looks good to upgrade as well

**Comment by vemv** [comment]
Nice fix, thanks much!

I will merge it. First, please:

* satisfy the small suggestion
* satisfy the PR checklist
* run `export LEIN_JVM_OPTS="-Dmranderson.internal.no-parallelism=true"; PROJECT_VERSION=0.45.0 make install` in the project root
  * make sure 0.45.0 matches with the version you normally use
  * use this locally-installed version for a reasonable period, so that we can gain confidence in that nothing in the debugger broke.

Cheers - V

File: src/cider/nrepl/middleware/debug.clj (hunk 1 above)
Comment (darkleaf) [reply src/cider/nrepl/middleware/debug.clj]: done!

**Comment by darkleaf** [comment]
Снимок экрана 2024-02-04 в 22 07 40

File: src/cider/nrepl/middleware/debug.clj [hunk 2]
```diff
@@ -570,9 +572,8 @@ this map (identified by a key), and will `dissoc` it afterwards."}
   `irrelevant-return-value-forms`."
   [&env form]
   (or (and (symbol? form)
-           (not (contains? &env form))
```
Comment (darkleaf) [inline src/cider/nrepl/middleware/debug.clj:573]: `(not (contains? &env form))` is handled inside `(resolve &env form)`

**Comment by bbatsov** [comment]
The changes look good to me. @vemv did you by any chance forgot to revisit this PR?

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 ## master (unreleased)
 
+* `middleware.debug`: use `resolve` with `&env` to fix debugging in case of a local shadows a var.
```
Comment (bbatsov) [inline CHANGELOG.md:5]: It'd be nice to add a "Bugs fixed" heading above.

File: src/cider/nrepl/middleware/debug.clj [hunk 4]
```diff
@@ -448,16 +448,18 @@ this map (identified by a key), and will `dissoc` it afterwards."}
...
-          m (meta v)]
-      ;; We do not go so far as to actually try to read the code of the function
-      ;; at this point, which is at macroexpansion time.
-      (and v
-           (safe-to-debug? (:ns m))
-           (not (:macro m))
-           (not (:inline m))))))
+  ([form]
+   (looks-step-innable? nil form))
+  ([&env form]
```
Comment (bbatsov) [inline src/cider/nrepl/middleware/debug.clj:453]: You should also document the `env` param in the docstring.

**Comment by vemv** [comment]
@darkleaf have you been testing out this changes locally in the meantime?

**Comment by darkleaf** [comment]
> @darkleaf have you been testing out this changes locally in the meantime?
Yes, I have. It [works](https://github.com/clojure-emacs/cider-nrepl/pull/847#issuecomment-1925868634).

**Comment by vemv** [comment]
Awesome - thank you!

PR LGTM after addressing @bbatsov 's feedback

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,10 @@
 
 ## master (unreleased)
 
+## Bugs fixed
+
+* `middleware.debug`: use `resolve` with `&env` to fix debugging in case of a local shadows a var.
```
Comment (vemv) [inline CHANGELOG.md:7]:
```suggestion
* [#846](https://github.com/clojure-emacs/cider-nrepl/issues/846): `middleware.debug`: use `resolve` with `&env`, preventing exceptions if a local shadows a var.
```

Comment (darkleaf) [reply CHANGELOG.md]: I've added link to the PR. Thanks.
But there were no exceptions, just a local shadowing.

**Comment by vemv** [comment]
Thanks much!

There should be a release within the next few days.

**Pull request merged**